src/librustc/ty/subst.rs

```diff
@@ -202,19 +179,25 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
     where FR: FnMut(&ty::RegionParameterDef, &[Kind<'tcx>]) -> &'tcx ty::Region,
           FT: FnMut(&ty::TypeParameterDef<'tcx>, &[Kind<'tcx>]) -> Ty<'tcx> {
         let defs = tcx.lookup_generics(def_id);
-        let mut substs = Vec::with_capacity(defs.count());
-
-        Substs::fill_item(&mut substs, tcx, defs, &mut mk_region, &mut mk_type);
-
-        Substs::new(tcx, substs)
+        let count = defs.count();
+        if count <= 8 {
```

Comment:
I don't know that this is overall expensive--it's one if branch. However, @nnethercote may be able to comment on how hot this function is. I personally am not quite sure what it does, so not sure how hot it could be.